subprojects/core/src/main/java/org/gradle/configuration/DefaultScriptPluginFactory.java

```diff
@@ -124,7 +130,7 @@ public void apply(final Object target) {
                     return fileLookup.getFileResolver().getPatternSetFactory();
                 }
             };
-            services.add(ScriptPluginFactory.class, DefaultScriptPluginFactory.this);
+            services.add(ScriptPluginFactory.class, scriptPluginFactory);
```

Comment:
I think you can completely break the cycle by making the selector not depend on the DefaultScriptPluginFactory. Instead, the DefaultScriptPluginFactory should live in the `pluginUse` project and be discovered using the same ServiceLocator pattern that the KotlinScriptPluginFactory uses.

Let's not keep this PR waiting, we can do that in a follow-up.